Minimize fetches from empty queues

sensorsasha opened the pull request:
### Overview

This PR addresses a flaw in the current fetching system. Currently, we have to run `rpoplpush` for every queue trying to fetch a job. 

https://github.com/sensortower/sidekiq-ultimate/blob/b43c117b939eb68521e12c3f7601c12942bb2631/lib/sidekiq/ultimate/fetch.rb#L64-L69

The more requests we need to perform before finding a queue with a job to pick up, the more time is wasted. If the system has a lot of high priority but most of the time empty queues, it's even more vulnerable. Every time we have to iterate over empty queues. 

[I run a benchmark to test how slow is sidekiq-ultimate fetching](https://gist.github.com/sensorsasha/8895c184d3a651fe4c6788e8f89ff7e4). Here is a comparison of regular fetch vs existing implementation based on the number of empty queues (X axis):
![232784246-2cf2cc67-6779-491e-ab65-7daad311755b](https://user-images.githubusercontent.com/50340647/233662874-fe5f632a-4788-4a5d-aea4-b3c7c5a01d07.png)

We do have a "queue exhaustion" mechanism
https://github.com/sensortower/sidekiq-ultimate/blob/b43c117b939eb68521e12c3f7601c12942bb2631/lib/sidekiq/ultimate/fetch.rb#L68

But it's per sidekiq thread. For bigger fleets of workers, it makes sense to share the list of empty queues. This is exactly what this PR is doing. 

### Implementation details

For each sidekiq **process** it adds a background `Concurrent::TimerTask`. It periodically (controllable by the `empty_queues_refresh_interval_sec` setting) tries to update both the global list of empty queues (stored in redis) and the local list (stored in a local variable).

Global list update is covered by a global lock (based on redis), so only a single process can update it at the same time.

Local list is stored in the `EmptyQueues` singleton which is accessible for reads by all the threads. 

Every call to `EmptyQueues.instance.refresh!` requires a local lock implemented using ruby `Mutex`.

### Bonus
- Fixes many flaky tests because of the shared state
- Fixes namespacing for redlock locks
- Adds tests for existing functionality in `lib/sidekiq/ultimate/fetch.rb`
- Use a cheaper check before trying to acquire a lock for the resurrection
- Tests for `Sidekiq::Ultimate.setup`

sensorsasha committed:
Minimize fetches from empty queues

sensorsasha committed:
Do not stop on the first CI fail

sensorsasha committed:
Fix redlock namespacing

sensorsasha committed:
Update docs

sensorsasha committed:
Update version

sensorsasha committed:
Update appraisals

sensorsasha committed:
Do not fail if no namespace used

sensorsasha committed:
Sidekiq prefixes queue names so do we

sensorsasha committed:
Logging for local list refresh

sensorsasha committed:
Extract RedisSscan

davidbodow-st commented on README.md:
````diff
@@ -86,6 +86,21 @@ Sidekiq::Ultimate::Resurrector::Count.read(:job_id => "2647c4fe13acc692326bd4c2"
+
+```ruby
+Sidekiq::Ultimate.setup! do |config|
+  config.empty_queues_refresh_interval = 42
````
Nit: When no unit is given, I assume the base SI unit, so not strictly necessary, but IMO it reduces confusion to append the unit anyway.
```suggestion
  config.empty_queues_refresh_interval_sec = 42
```

sensorsasha replied:
Updated.

davidbodow-st commented on README.md:
```diff
@@ -86,6 +86,21 @@ Sidekiq::Ultimate::Resurrector::Count.read(:job_id => "2647c4fe13acc692326bd4c2"
+the moment when that new job is picked up.
+
+**Note** that every worker needs to maintain its own local list of empty queues. Setting this interval to a low
+values will increase the number of redis calls and will increase the load on redis.
```
```suggestion
Specifies how often the list of empty queues should be refreshed.
In a nutshell, this sets the maximum possible delay between when a job was pushed to previously empty queue and earliest the moment when that new job could be picked up.

**Note:** every worker maintains its own local list of empty queues.
Setting this interval to a low value will increase the number of Redis calls needed to check for empty queues, increasing the total load on Redis.

This setting helps manage the tradeoff between performance penalties and latency needed for reliable fetch.
Under the hood, Sidekiq's default fetch occurs with [a single Redis `BRPOP` call](https://redis.io/commands/brpop/) which is passes list of all queues to pluck work from.
In contrast, [reliable fetch uses `LPOPRPUSH`](https://redis.io/commands/rpoplpush/) (or the equivalent `LMOVE` in later Redis versions) to place in progress work into a WIP queue.
However, `LPOPRPUSH` can only check one source queue to pop from at once, and [no multi-key alternative is available](https://github.com/redis/redis/issues/1785), so multiple Redis calls are needed to pluck work if an empty queue is checked.
In order to avoid performance penalties for repeated calls to empty queues, Sidekiq Ultimate therefore maintains a list of recently know empty queues which it will avoid polling for work.

Therefore:
- If your Sidekiq architecture has *a low number of total queues*, the worst case penalty for polling empty queues will be bounded, and it is reasonable to **set a shorter refresh period**.
- If your Sidekiq architecture has a *high number of total queues*, the worst case penalty for polling empty queues is large, and it is recommended to **set a longer refresh period**.
- When adjusting this setting:
    - Check that work is consumed appropriately quickly from high priority queues after they bottom out (after increasing the refresh interval)
    - Check that backlog work does not accumulate in low priority queues (after decreasing the refresh interval)

```

davidbodow-st commented on lib/sidekiq/ultimate/configuration.rb:
```diff
@@ -16,7 +16,36 @@ class Configuration
+          raise ArgumentError, "Invalid 'empty_queues_refresh_interval' value: #{value}. Must be Numeric"
+        end
+
+        @empty_queues_refresh_interval = value.to_i
```
You're already checking it's numeric, so why not just check it's an integer? Or, just support floats?

sensorsasha replied:
👍 `.to_i` is not needed here at all.

davidbodow-st commented on lib/sidekiq/ultimate/empty_queues.rb:
```diff
@@ -0,0 +1,155 @@
+            :run_now            => true,
+            :execution_interval => Sidekiq::Ultimate::Configuration.instance.empty_queues_refresh_interval
+          }) { Sidekiq::Ultimate::EmptyQueues.instance.refresh! }
+        end
```
```suggestion
        Sidekiq.on(:startup) do
          refresher&.shutdown

          jitter = 0.1 * Sidekiq::Ultimate::Configuration.instance.empty_queues_refresh_interval
          offset = rand(2 * jitter) - jitter
          
          refresher = Concurrent::TimerTask.execute({
            :run_now            => true,
            :execution_interval => Sidekiq::Ultimate::Configuration.instance.empty_queues_refresh_interval + offset
          }) { Sidekiq::Ultimate::EmptyQueues.instance.refresh! }
        end
```

davidbodow-st replied:
Also add jitter for other lock intervals

sensorsasha replied:
Added jitter to all 3 timer tasks.

davidbodow-st commented on lib/sidekiq/ultimate/fetch.rb:
```diff
@@ -55,37 +52,36 @@ def self.bulk_requeue(units, _options)
+        queues = (@strict ? @queues : @queues.shuffle.uniq) - @exhausted_by_throttling.to_a - @empty_queues.queues
 
         # Avoid calling heavier `paused_queue` if there's nothing to filter out
         return queues if queues.empty?
```
Missing the map below

sensorsasha replied:
Oh, it doesn't matter since `queues` is empty here 🙊

davidbodow-st commented on spec/sidekiq/ultimate/configuration_spec.rb:
```diff
@@ -0,0 +1,24 @@
+  describe "empty_queues_refresh_interval" do
+    let(:instance) { Class.new(described_class).instance }
+
+    it "defaults to 60 seconds" do
```
30?

davidbodow-st commented on spec/sidekiq/ultimate/empty_queues_spec.rb:
```diff
@@ -0,0 +1,127 @@
+      described_class.setup!
+
+      expect { sidekiq_util.fire_event(:startup) }.
+        to change { ObjectSpace.each_object(Concurrent::TimerTask).count(&:running?) }.from(0).to(1)
```
Can check for the exact name for the timer tasks maybe?

sensorsasha replied:
Updated to use a separate class. E.g.

```ruby
TASK_CLASS = Class.new(Concurrent::TimerTask)
```

davidbodow-st reviewed (commented):
Ideas for next iteration if performance is still too slow:
- Avoid stale empty queue lists by checking the global list more quickly if we tried to read during a lock
- If a queue bottoms out during a cycle, add it to the lists for the rest of the interval so that we avoid double-hitting a known empty queue

sensorsasha committed:
Update README.md

Co-authored-by: David Bodow <david.bodow@sensortower.com>

sensorsasha committed:
- rename empty_queues_refresh_interval to empty_queues_refresh_interval_sec
- add jitter for refresh interval

sensorsasha committed:
Add jitter to resurrector timers

stefanl-st commented:
I think we missed a *pretty important* reviewer here. @ixti please take a look if you wish

stefanl-st commented on lib/sidekiq/ultimate/configuration.rb:
```diff
@@ -16,7 +16,36 @@ class Configuration
+      # If `enable_resurrection_counter` setting is enabled, on each resurrection event, a counter is increased.
+      # Counter value is stored in redis by jid and has expiration time 24 hours.
+      # @return [Boolean]
+      attr_accessor :enable_resurrection_counter
```
```suggestion
      # If `enable_resurrection_counter` setting is enabled, on each resurrection event, a counter is increased.
      # This is useful for telemetry purposes in order to understand how often jobs are resurrected
      # Counter value is stored in redis by jid and has expiration time 24 hours.
      # @return [Boolean]
      attr_accessor :enable_resurrection_counter
```

stefanl-st commented on README.md:
````diff
@@ -86,6 +86,34 @@ Sidekiq::Ultimate::Resurrector::Count.read(:job_id => "2647c4fe13acc692326bd4c2"
 => 1
 ```
 
+### Empty queues list refresh interval
````
```suggestion
### Empty Queues List Refresh Interval
```

stefanl-st commented on lib/sidekiq/ultimate/configuration.rb:
```diff
@@ -16,7 +16,36 @@ class Configuration
+      # Note that every worker needs to maintain its own local list of empty queues. Setting this interval to a low
+      # values will increase the number of redis calls and will increase the load on redis.
+      # @return [Integer] interval in seconds to refresh the list of empty queues
+      attr_reader :empty_queues_refresh_interval_sec
```
Think we can do better in terms of naming. Lets specify that it's a cache and that is global (to differentiate from the individual worker cache)

```suggestion
      attr_reader :global_empty_queue_cache_refresh_interval_sec
```

sensorsasha replied:
This is not just global, it's both global and local. I like the "cache" part tho.

stefanl-st commented on lib/sidekiq/ultimate/empty_queues.rb:
```diff
@@ -0,0 +1,152 @@
+    # Maintains a list of empty queues. It has a global list and a local list.
+    # The global list is stored in redis and updated periodically. The local list is updated either by using the fresh
+    # list fetched for global list update or by using existing global list.
+    # Only one process can update the global list at a time.
```
process? worker? actor? wording is important.

sensorsasha replied:
I meant exactly process. Btw this helpful regarding sidekiq naming https://github.com/sidekiq/sidekiq/issues/4955

stefanl-st commented on lib/sidekiq/ultimate/empty_queues.rb:
```diff
@@ -0,0 +1,152 @@
+      # It will call #refresh! every
+      # `Sidekiq::Ultimate::Configuration.instance.empty_queues_refresh_interval_sec` seconds
+      def self.setup!
+        refresher = nil
```
```suggestion
        refresher_timer_task = nil
```

stefanl-st commented on lib/sidekiq/ultimate/empty_queues.rb:
```diff
@@ -0,0 +1,152 @@
+        return false unless local_lock.try_lock
+
+        begin
+          refresh_global_list! || refresh_local_list!
```
Nice!

davidbodow-st commented on spec/spec_helper.rb:
```diff
@@ -3,6 +3,9 @@
 require_relative "support/simplecov"
 require "sidekiq"
 
+# Silence ruby warnings
+$TESTING = true # rubocop:disable Style/GlobalVars
```
Hmm, I actually think it's helpful to see them, especially if we need to deal with deprecation warnings for instance

ixti replied:
$TESTING is a sidekiq global variable. It does not silence Ruby warnings, when ruby executed with `-w` argument or `RUBYOPT=-w` environment variable.

See Ruby globals: https://docs.ruby-lang.org/en/2.7.0/globals_rdoc.html

sensorsasha replied:
This is a different kind of warnings:

```
RUBYOPT=-W0 rspec
gems/sidekiq-5.2.10/lib/sidekiq/processor.rb:66: warning: global variable `$TESTING' not initialized
gems/sidekiq-5.2.10/lib/sidekiq/launcher.rb:63: warning: global variable `$TESTING' not initialized
gems/sidekiq-5.2.10/lib/sidekiq/cli.rb:17: warning: global variable `$TESTING' not initialized
```

stefanl-st commented on lib/sidekiq/ultimate/fetch.rb:
```diff
@@ -14,17 +15,14 @@ class Fetch
       def initialize(options)
-        @exhausted = ExpirableSet.new
-
+        @exhausted_by_throttling = ExpirableSet.new
```
Thank you

stefanl-st commented on lib/sidekiq/ultimate/fetch.rb:
```diff
@@ -14,17 +15,14 @@ class Fetch
-      QUEUE_TIMEOUT = 5
-
       # Delay between queue poll attempts if it's last job was throttled.
       THROTTLE_TIMEOUT = 15
```
I think we should try to change this to a lower value. Maybe 5?

sensorsasha replied:
I'll extract it into configuration.

sensorsasha replied:
Extracted.

stefanl-st commented on lib/sidekiq/ultimate/fetch.rb:
```diff
@@ -55,37 +52,36 @@ def self.bulk_requeue(units, _options)
 
         # Avoid calling heavier `paused_queue` if there's nothing to filter out
-        return queues if queues.empty?
+        return [] if queues.empty?
```
❤️

stefanl-st commented on lib/sidekiq/ultimate/interval_with_jitter.rb:
```diff
@@ -0,0 +1,19 @@
+        # Jitter is +- RANDOM_OFFSET_RATION from the original value.
+        def call(interval)
+          jitter_ratio = (rand(0..RANDOM_OFFSET_RATIO) * 2) - RANDOM_OFFSET_RATIO
+          interval + (jitter_ratio * interval)
```
Probably cleaner

```suggestion
          jitter_factor = 1 + rand(-RANDOM_OFFSET_RATIO..RANDOM_OFFSET_RATIO) 
          jitter_factor * interval
```

stefanl-st commented on lib/sidekiq/ultimate/redis_sscan.rb:
```diff
@@ -0,0 +1,23 @@
+            result.push(*values)
+            break if cursor == "0"
+          end
+          result.uniq! # Cursor is not atomic, so there may be duplicates because of concurrent update operations
+          result
```
Why does `#uniq!` return `nil`? 😢

ixti replied:
Most bang-ending Ruby operators return `nil` if there were no changes. That's done to allow do something when changes were made or not:

``` ruby
if arr.uniq!
  puts "we removed some duplicates"
end
```

stefanl-st replied:
Makes sense. Thank you.

davidbodow-st replied:
> Most bang-ending Ruby operators

It's unfortunately not a very well followed standard:

```
irb(main):009:0> a = %w[x y z a b c c d d e]
=> ["x", "y", "z", "a", "b", "c", "c", "d", "d", "e"]
irb(main):010:0> a.sort!
=> ["a", "b", "c", "c", "d", "d", "e", "x", "y", "z"]
irb(main):011:0> a.sort!
=> ["a", "b", "c", "c", "d", "d", "e", "x", "y", "z"]
irb(main):012:0> a.uniq!
=> ["a", "b", "c", "d", "e", "x", "y", "z"]
irb(main):013:0> a.uniq!
=> nil
```

Also: `#save!` in ActiveRecord...

The root of the issue is that bang methods are meant to be "not ugly" and just indicate that a method is more dangerous than an alternative non-bang method, [per Matz](https://www.ruby-forum.com/t/conventions-in-ruby-and-the-principle-of-least-surprise/158706/2).  Contrast with various Lisp dialect conventions where it means "this causes a side effect" | [example](https://en.wikibooks.org/wiki/Scheme_Programming/Mutability). One of the worse Ruby aesthetics IMO because it ends up that nothing is predictable, and plenty of people leave bang methods around the code anyway without a non-bang "safer" alternative, so the original goal isn't achieved either.

/rant :wink:

stefanl-st commented on lib/sidekiq/ultimate/resurrector.rb:
```diff
@@ -19,9 +21,8 @@ module Resurrector
-      # Redis-rb 4.2.0 renamed `#exists` to `#exists?` and changed behaviour of `#exists` to return integer
-      # https://github.com/redis/redis-rb/blob/master/CHANGELOG.md#420
-      USE_EXISTS_QUESTION_MARK = Gem::Version.new(Redis::VERSION) >= Gem::Version.new("4.2.0")
+      CtulhuTimerTask = Class.new(Concurrent::TimerTask)
+      AedTimerTask = Class.new(Concurrent::TimerTask)
```
It's funny sure, but maybe more descriptive names would be better? 🐙

kevinrobell-st replied:
Or at least some sort of explanation or background would be nice.

sensorsasha replied:
I'll fix it as a follow up.

sensorsasha replied:
I fixed it in the scope of this PR.

stefanl-st commented on lib/sidekiq/ultimate/resurrector/lock.rb:
```diff
@@ -20,20 +20,32 @@ class Lock
+          end
+
+          def resurrected_recently?(redis)
+            results  = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
+            distance = results[0][0] - results[1].to_i
```
Can't you write something like

```suggestion
            start, end  = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
            distance = end[0] - start.to_i
```

stefanl-st replied:
And why is only one of these `.to_i`?

ixti replied:
`TIME` is guaranteed to return two integers, `GET` on the other hand, can return just about anything. In this case we trust that `LAST_RUN_KEY` stores `Integer`, but it might not be set yet, in this case it will return `nil`

ixti replied:
BTW, there's a bit more fancy way to write this:

``` ruby
(curr, *), last = redis.pipelined { |pipeline| [pipeline.time, pipeline.get(LAST_RUN_KEY)] }
distance = curr - last.to_i
```

Furthermore, I think it's better to use `MULTI` instead of `PIPELINE`.

stefanl-st commented on lib/sidekiq/ultimate/use_exists_question_mark.rb:
```diff
@@ -0,0 +1,17 @@
+  module Ultimate
+    # Redis-rb 4.2.0 renamed `#exists` to `#exists?` and changed behaviour of `#exists` to return integer
+    # https://github.com/redis/redis-rb/blob/master/CHANGELOG.md#420
+    module UseExistsQuestionMark
+      def self.use?
+        return @use if defined?(@use)
+
+        @use = Gem::Version.new(Redis::VERSION) >= Gem::Version.new("4.2.0")
+      end
+    end
```
Why not just make a direct convenience function? `Sidekiq::Ultimate.redis_exists?(r, key)`

ixti replied:
I suggest create a whole new class:

``` ruby
module Sidekiq
  module Ultimate
    class RedisAdapter
      REDIS_420 = Gem::Version.new("4.2.0").freeze

      def initialize(redis)
        @redis = redis
      end

      def exists?(key)
        if REDIS_420 <= @redis.class::VERSION
          @redis.exists?(key)
        else
          @redis.exists(key)
      end

      # more method here
    end
  end
end
```

This way it will be much easier to support Sidekiq 7.0, which uses redis-client instead of redis-rb.

ixti replied:
Actually there's no need to check for redis client version:

``` ruby
def redis_exists?(redis, key)
  redis.exists(key).then do |value|
    # redis < 4.2:  exists(key) => true|false
    # redis >= 4.2: exists(key) => 1|0
    value == 1 || value == true
  end
end
```

sensorsasha replied:
I'm going to drop support for redis < 4.2 soon, so decided to keep it simple.

stefanl-st commented on README.md:
````diff
@@ -86,6 +86,34 @@ Sidekiq::Ultimate::Resurrector::Count.read(:job_id => "2647c4fe13acc692326bd4c2"
+end
+```
+
+Specifies how often the list of empty queues should be refreshed.
+In a nutshell, this sets the maximum possible delay between when a job was pushed to previously empty queue and earliest the moment when that new job could be picked up.
+
+**Note:** every worker maintains its own local list of empty queues.
+Setting this interval to a low value will increase the number of Redis calls needed to check for empty queues, increasing the total load on Redis.
+
+This setting helps manage the tradeoff between performance penalties and latency needed for reliable fetch.
+Under the hood, Sidekiq's default fetch occurs with [a single Redis `BRPOP` call](https://redis.io/commands/brpop/) which is passes list of all queues to pluck work from.
+In contrast, [reliable fetch uses `LPOPRPUSH`](https://redis.io/commands/rpoplpush/) (or the equivalent `LMOVE` in later Redis versions) to place in progress work into a WIP queue.
+However, `LPOPRPUSH` can only check one source queue to pop from at once, and [no multi-key alternative is available](https://github.com/redis/redis/issues/1785), so multiple Redis calls are needed to pluck work if an empty queue is checked.
+In order to avoid performance penalties for repeated calls to empty queues, Sidekiq Ultimate therefore maintains a list of recently know empty queues which it will avoid polling for work.
+
+Therefore:
+- If your Sidekiq architecture has *a low number of total queues*, the worst case penalty for polling empty queues will be bounded, and it is reasonable to **set a shorter refresh period**.
+- If your Sidekiq architecture has a *high number of total queues*, the worst case penalty for polling empty queues is large, and it is recommended to **set a longer refresh period**.
+- When adjusting this setting:
+    - Check that work is consumed appropriately quickly from high priority queues after they bottom out (after increasing the refresh interval)
+    - Check that backlog work does not accumulate in low priority queues (after decreasing the refresh interval)
````
I really like that you added all this detail but in order to make it a bit digestible, please rewrite it as suggested below:

- Start by explaining the problem this feature solves
- Explain how it works under the hood (the different types of Redis operations like you do now)
- Explain what the performance implications are for a shared global queue (also please include the total number of workers into the description, it looks like its missing)
- Then explain the constant itself

sensorsasha replied:
@davidbodow-st any thoughts since you proposed this section?

davidbodow-st replied:
@stefanl-st @sensorsasha TBH, I just wrote this off the top of my head, so we can sink more time into refining the structure and wording if needed, though I think it's probably at diminishing returns right now.

I also disagree with the proposed structure, as IMHO, it would provide too much detail for quick reference use cases, which are the majority of documentation readers' time. Here, I assume that most readers will need to learn the "why" just once and the "what was that config option called again?" many times while using the feature.

So, my preferred structure is:
- Quick reference overview
- Explain the problem that needs to be solved
- Explain the tradeoff that the implementation details require us to make

IMO we're already following this structure, and it's mostly optimized for the quick reference use case. If we want to take another pass, then I would recommend:
- Tighten up the "Therefore" section and move it into the quick reference, for readers who care about making the correct decision without fully building the mental model of the implementation that they need to do that
- Take a general pass to make the language more concise (could definitely be condensed)

@stefanl-st If you disagree / would still prefer your structure, can you elaborate on which audience you want to optimize for instead of quick reference readers (who would expect the `Then explain the constant itself` part at the very beginning, IMO)?

stefanl-st commented on lib/sidekiq/ultimate/configuration.rb:
```diff
@@ -16,7 +16,36 @@ class Configuration
+      # @return [Boolean]
+      attr_accessor :enable_resurrection_counter
+
+      # It specifies how often the list of empty queues should be refreshed.
+      # In a nutshell, it specifies the maximum possible delay between a job was pushed to previously empty queue and
+      # the moment when that new job is picked up.
+      # Note that every worker needs to maintain its own local list of empty queues. Setting this interval to a low
+      # values will increase the number of redis calls and will increase the load on redis.
+      # @return [Integer] interval in seconds to refresh the list of empty queues
```
Defer some of this explanation to the readme and link it.

```suggestion
      # Each individual worker ignores attempting to fetch jobs from queues it believes are empty by 
      # checking the empty queue cached state from the global empty queue cache to speed up performance
      # @return [Integer] interval in seconds between global cache refreshes
```

sensorsasha replied:
I'd rather keep it here and link README to these comments. IMO it's easier to keep it up to date than README.

davidbodow-st replied:
I'd also be fine with that to tighten up the Readme language

stefanl-st commented on lib/sidekiq/ultimate/empty_queues.rb:
```diff
@@ -0,0 +1,152 @@
+        Sidekiq.on(:shutdown) { refresher&.shutdown }
+      end
+
+      # It updates the global list of empty queues and the local list of empty queues.
+      # During the update, the global list is locked to prevent other processes from updating it since this operation
+      # is expensive.
+      # The local list is updated anyway. Either by using the fresh list fetched for global list update or by using
+      # existing global list.
+      #
+      # @return [Boolean] true if local list was updated
```
```suggestion
      # Attempts to update the global cache of empty queues by first acquiring a global lock
      # If the lock is acquired, it brute force generates an accurate list of currently empty queues and then writes this 
      # updated list to the global cache
      # The local queue cache is always updated as a result of this operation, either by using the recently generated 
      # list or fetching the most recent list from the global cache
      #
      # @return [Boolean] true if local list was updated
```

stefanl-st commented on lib/sidekiq/ultimate/empty_queues.rb:
```diff
@@ -0,0 +1,152 @@
+        end
+      end
+
+      def fetch_empty_queues(redis)
```
```suggestion
      def generate_empty_queues(redis)
```

stefanl-st reviewed (approved):
This is great @sensorsasha, overall the direction is solid and I think this will be a significant improvement.

On the top level after reading the code I suggest we change the code and wording to treat this more like a cache and therefore use words that are commonly used with caches so use the words *global* and *local cache* rather than using the words *global* and *local list*. 

An example of the change is:

`#set_local_list` `=>` `#update_local_cache`

There are many places where need to make changes and it also depends on the context so if you think its a good idea, please take a stab at it and I can review after.

ixti commented on lib/sidekiq/ultimate/configuration.rb:
```diff
@@ -16,7 +16,36 @@ class Configuration
+      end
+
+      def empty_queues_refresh_interval_sec=(value)
+        unless value.is_a?(Numeric)
```
Attribute has Integer type. Probably worth to require Integer here as well, as Floats are also Numerics.

sensorsasha replied:
Well, this seems to be a mistake in the concurrent-ruby docs but the float is actually supported. I've checked the code and tested it manually before adding this check here 😄 

`ScheduledTask` is used by `TimerTask` under the hood: 
https://github.com/ruby-concurrency/concurrent-ruby/blob/e7ff945661a93a2e094b3259f23ccb1b3eea9bcf/lib/concurrent-ruby/concurrent/scheduled_task.rb#L167


```ruby
Concurrent::TimerTask.execute({:execution_interval => 0.1}) { puts Time.now.to_f }

1682345298.779785
1682345298.881609
1682345298.9855418
1682345299.091567
1682345299.1970491
1682345299.302599
1682345299.407186
1682345299.510379
1682345299.6152
1682345299.720865
1682345299.826708
1682345299.933086
1682345300.0387928
1682345300.144703
1682345300.2506618
1682345300.353
1682345300.4590552
```

I've [opened a PR](https://github.com/ruby-concurrency/concurrent-ruby/pull/994) to fix the docs.

ixti replied:
We should change the documentation of the attr_reader then too.

sensorsasha replied:
Done 👍

ixti commented on lib/sidekiq/ultimate/redis_sscan.rb:
```diff
@@ -0,0 +1,23 @@
+            result.push(*values)
+            break if cursor == "0"
+          end
+          result.uniq! # Cursor is not atomic, so there may be duplicates because of concurrent update operations
```
```suggestion
          # SSCAN is not atomic, so there may be duplicates because of concurrent update operations
          # See: https://redis.io/commands/scan/#scan-guarantees
          redis.sscan_each(key).to_a.uniq
```

sensorsasha replied:
I just removed this class entirely.

ixti reviewed (approved):
LGTM, but IMO there's no need in re-implementing sscan_each. Both redis-rb and redis-client have convenience wrappers for that build-in.

kevinrobell-st commented on lib/sidekiq/ultimate/interval_with_jitter.rb:
```diff
@@ -0,0 +1,19 @@
+
+      class << self
+        # Returns execution interval with jitter.
+        # Jitter is +- RANDOM_OFFSET_RATION from the original value.
```
```suggestion
        # Jitter is +- RANDOM_OFFSET_RATIO from the original value.
```
Nit: typo

kevinrobell-st reviewed (approved):
I agree with a lot of the comments about documentation but, in general, I think this approach makes sense.

The only thing in the back of my mind is whether 30 seconds is a sane default for the `empty_queues_refresh_interval_sec`. Do we know how long queues usually stay empty for?

stefanl-st commented on lib/sidekiq/ultimate/resurrector.rb:
```diff
@@ -19,9 +21,8 @@ module Resurrector
-      # Redis-rb 4.2.0 renamed `#exists` to `#exists?` and changed behaviour of `#exists` to return integer
-      # https://github.com/redis/redis-rb/blob/master/CHANGELOG.md#420
-      USE_EXISTS_QUESTION_MARK = Gem::Version.new(Redis::VERSION) >= Gem::Version.new("4.2.0")
+      CtulhuTimerTask = Class.new(Concurrent::TimerTask)
```
[Typo ](https://en.wikipedia.org/wiki/Cthulhu#:~:text=Cthulhu%20is%20a%20fictional%20cosmic,magazine%20Weird%20Tales%20in%201928.)

sensorsasha committed:
Apply suggestions from code review

Co-authored-by: Stefan Lynggaard <stefan@sensortower.com>
Co-authored-by: Kevin Robell <kevin.robell@sensortower.com>

sensorsasha committed:
- List -> Cache
- Add configuration for THROTTLE_TIMEOUT
- Address proposed naming changes
- Remove RedisSscan in favor of `.sscan_each` from redis-rb
- aed -> heartbeat
- cthulhu -> resurrect

sensorsasha commented:
> LGTM, but IMO there's no need in re-implementing sscan_each. Both redis-rb and redis-client have convenience wrappers for that build-in.

Thank you for pointing this out. I did a brief check of redis-rb docs but missed that for some reason. 🤝

sensorsasha commented:
> The only thing in the back of my mind is whether 30 seconds is a sane default for the `empty_queues_refresh_interval_sec`. Do we know how long queues usually stay empty for?

Not really. I'd say that 30 seconds looks safe to me. The rest can be tuned for any specific use case.

sensorsasha committed:
Fix type in the docs of empty_queues_cache_refresh_interval_sec

sensorsasha commented:
> #set_local_list => #update_local_cache

I like this idea. It definitely makes it easier to reason about.I updated the code. @stefanl-st
